### ISSUE TITLE
Reproducibility Issue

### ISSUE THREAD

**Issue by lthilnklover** [SMLC-NYSBC/TARGET-VAE#2]
Hi,

I am currently trying to reproduce the MNIST-U, MNIST-N clustering experiments. However, for some reason, I am not able to reproduce the numbers in your paper. Some of the possible reasons are follows:

**1) Dataset** : Although you have kindly provided MNIST-U, MNIST-N datasets, those datasets lack the labels for evaluating the clustering accuracy. Hence I have created my own datasets (using your configuration). Maybe this could be the reason for the discrepency.

**2) Hyperparameters** (including architecture of the model) : In your paper, you have not explicitly provided the detailed hyperparameters for MNIST-U, MNIST-N experiments (sorry in advance if I have missed something). So I was guessing the default configuration you provided in the README, is the setting you used in the paper. Maybe my guess was wrong.

Or, possibily, the difference is due to some other reasons. Hence I would be grateful if you could provide the detailed configuration for MNIST-U and MNIST-N experiments. 

Sincerely,

Joo Young Choi

**Comment by alireza-nasiri** [comment]
Hi, 
I think it is very likely that this issue is happening because of the discrepancy in the labels. For MNIST, we use the dataset from torchvision which also contains the semantic labels and works with the clustering_mnist.py file. You can get this mnist by these commands:
mnist_train = torchvision.datasets.MNIST('data/', train=True, download=True)
mnist_test = torchvision.datasets.MNIST('data/', train=False, download=True)

The command from Readme file, should give you the results for z_dim=2 and P8. However, for higher clustering accuracy, you can increase the z-dim to 32 by using –z-dim 32, and for better rotation prediction, you can use –groupconv 16.

Best,
Ali Nasiri

**Comment by lthilnklover** [comment]
Thank you for the reply. Just to clarity some points,

1) I have been using z_dim = 32 groupconv=16.

2) I am little confused, but are you saying when clustering, you used mnist (without rotation and translation)? In my case, I used your dataset (MNIST-N and MNIST-U) and used the labels from MNIST provided by torchvision. (For testset, the order of the samples were same for your dataset and torchvision dataset).

However, according to my experiments, the clustering accuracy seems to vary a lot depending on the run. For MNIST-U, the average was around 67% with maximum of 72% and mimimum of 61%. Did you get the similar statistics? or am I doing something wrong. 

command: 

`CUDA_VISIBLE_DEVICES=0 python train_mnist.py -z 32 --dataset mnist-U --t-inf attention --r-inf attention+offsets --groupconv 16 --fourier-expansion --log-root /extdata1/lthilnklover/vae/target_vae_original`

y_labels code:


```
y_labels = torchvision.datasets.MNIST('/extdata1/lthilnklover/data/', train=False, download=True)

array = torch.zeros((len(mnist_test),), dtype=torch.uint8)
for i in range(len(mnist_test)):
    array[i] = y_labels[i][1]
y_labels = array
```

**Comment by alireza-nasiri** [comment]
1. We get our best results on MNIST-U with z-dim=32 and groupconv=16, so your setting for these values are correct.
2.  For clustering MNIST, you can use the labels from the MNIST provided by torchvision. I was saying the same thing.

The reason for the high variance in the results is using agglomerative clustering. You can use k-means instead by adding --clustering k-means to your clustering command. I modified the code so now the k-means is the default clustering method  for MNIST.
Also I modified the code for RandomFourierEmbedding2d in the src/models.py, I modified the code so that the weight and bias won't change during the training which could have been another source of the variation in the clusterin result. Please use this new code for training the model.
 There is about 2 - 3 percent variation in the clustering result for the mnist, but it shouldn't be higher than that.
